### PR TITLE
Provide docs for order-granted-refund

### PR DESCRIPTION
**Pull request by korycins**
Add a chapter related to `OrderGrantedRefund`. 
It also includes the scope of changes described in RFC: https://github.com/saleor/saleor/discussions/15458
The PR with mentioned changes: https://github.com/saleor/saleor/pull/15590

### CODE REVIEW

File: docs/developer/payments.mdx [hunk 1]
```diff
@@ -852,6 +852,199 @@ sequenceDiagram
...
+There are two ways to process refunds:
+
+- Manual refund: This method involves directly refunding the payment through the payment app.
+- Refund based on `OrderGrantedRefund`: This method triggers the refund request based on the details stored in `OrderGrantedRefund`.
+
+#### Manual refund
+
+Manual refund can be triggered by calling `transactionRequestAction`. If the refund is successful, the `transaction.chargedAmount` will be reduced. The order's `authorizeStatus`, `chargeStatus`, and `totalBalance` will also be recalculated based on the new values of `chargedAmount` and `refundedAmount`. This method is useful when handling overcharged orders.
+
+Below example shows how to trigger manual refund:
```
Comment (2can) [inline docs/developer/payments.mdx:866]:
```suggestion
The below example shows how to trigger the manual refund:
```

File: docs/developer/payments.mdx [hunk 2]
```diff
@@ -852,6 +852,199 @@ sequenceDiagram
...
+### Processing refunds for Order
+
+There are two ways to process refunds:
+
+- Manual refund: This method involves directly refunding the payment through the payment app.
+- Refund based on `OrderGrantedRefund`: This method triggers the refund request based on the details stored in `OrderGrantedRefund`.
+
+#### Manual refund
+
+Manual refund can be triggered by calling `transactionRequestAction`. If the refund is successful, the `transaction.chargedAmount` will be reduced. The order's `authorizeStatus`, `chargeStatus`, and `totalBalance` will also be recalculated based on the new values of `chargedAmount` and `refundedAmount`. This method is useful when handling overcharged orders.
```
Comment (2can) [inline docs/developer/payments.mdx:864]:
```suggestion
The manual refund can be triggered by calling `transactionRequestAction`. If the refund is successful, the `transaction.chargedAmount` will be reduced. The order's `authorizeStatus`, `chargeStatus`, and `totalBalance` will also be recalculated based on the new values of `chargedAmount` and `refundedAmount`. This method is useful when handling overcharged orders.
```

File: docs/developer/payments.mdx [hunk 3]
```diff
@@ -852,6 +852,199 @@ sequenceDiagram
...
+
+The `OrderGrantedRefund` has an impact on the `authorizeStatus`, `chargeStatus`, and `totalBalance`, as it reduces the total value used to calculate the `totalBalance`. For example, if an order has a total of 100 USD and a single `TransactionItem` with a `chargedAmount` of 100 USD, the `totalBalance` would be 0, as `chargedAmount` - `order.total` results in 0. Adding a granted refund with an amount of 10 USD would result in a `totalBalance` of 10, as `chargedAmount` - (`order.total` - `grantedRefund.amount`) gives 10. The `chargeStatus` will be `OVERCHARGED`. If a refund request is made based on the defined granted refund and is successfully processed by the payment app, it will reduce the `chargedAmount`. The `totalBalance` will be 0, as `chargedAmount` (90USD) - (`order.total` - `grantedRefund.amount`) gives 0. The `chargeStatus` will be changed to `FULL`.
+
+The following example shows how to create the granted refund with assigned `TransactionItem` for the order.
```
Comment (2can) [inline docs/developer/payments.mdx:913]:
```suggestion
The following example shows how to create the granted refund with the assigned `TransactionItem` for the order.
```

File: docs/developer/payments.mdx [hunk 4]
```diff
@@ -852,6 +852,199 @@ sequenceDiagram
...
+- `input`:
+  - `lines` - List of lines related to the planned refund action:
+    - `id` - ID of the `OrderLine`.
+    - `quantity` - The quantity of the specific lines planned to refund.
+    - `reason` - Reason of the refund related to the specific line.
+  - `grantRefundForShipping` - Determines if the shipping costs will be also included in the refund.
+  - `amount` - Amount of the granted refund. If not provided, the amount will be calculated automatically based on provided `lines` and `grantRefundForShipping`.
+  - `transactionId` - ID of `TransactionItem` that will be used to process the refund. If `amount` is provided in the input, the `transaction.chargedAmount` needs to be equal to or greater than the provided `amount`. If `amount` is not provided in the input and calculated automatically by Saleor, the `min(calculatedAmount, transaction.chargedAmount)` will be used. This field was added in Saleor 3.20, and it will be a mandatory input field starting from Saleor 3.21.
+
+Below example shows how to update existing granted refund
```
Comment (2can) [inline docs/developer/payments.mdx:970]:
```suggestion
The below example shows how to update the existing granted refund
```

File: docs/developer/payments.mdx [hunk 5]
```diff
@@ -852,6 +852,199 @@ sequenceDiagram
...
+      field
+      message
+      code
+    }
+  }
+}
+```
+
+To request a refund, you need to provide the `grantedRefundId` field. The mandatory refund details are stored as `OrderGrantedRefund`, so the refund request will be created based on this data.
+The `OrderGrantedRefund.status` field will be updated based on the result of the refund action. The `TransactionEvent`s related to requested action will be accessible via `OrderGrantedRefund.transactionEvents` field.
```
Comment (2can) [inline docs/developer/payments.mdx:1046]:
```suggestion
The `OrderGrantedRefund.status` field will be updated based on the result of the refund action. The `TransactionEvent`s related to the requested action will be accessible via `OrderGrantedRefund.transactionEvents` field.
```

File: docs/developer/payments.mdx [hunk 6]
```diff
@@ -852,6 +852,199 @@ sequenceDiagram
...
+
+The mutation accepts below arguments:
+
+- `id` - ID of the `OrderGrantedRefund` which should be updated
+- `input`:
+  - `addLines`: Lines that should be added to `OrderGrantedrefund`
+    - `id` - ID of the `OrderLine`.
+    - `quantity` - The quantity of the specific lines planned to refund.
+    - `reason` - Reason of the planned refund related to the specific line.
+  - `removeLines` - List of `OrderGrantedRefundLine`'s IDs which should be removed from `OrderGrantedRefund`.
```
Comment (2can) [inline docs/developer/payments.mdx:1015]:
```suggestion
  - `removeLines` - List of `OrderGrantedRefundLine`'s IDs that should be removed from `OrderGrantedRefund`.
```

File: docs/developer/payments.mdx [hunk 7]
```diff
@@ -852,6 +852,199 @@ sequenceDiagram
...
+```
+
+The mutation accepts below arguments:
+
+- `id` - ID of the `OrderGrantedRefund` which should be updated
+- `input`:
+  - `addLines`: Lines that should be added to `OrderGrantedrefund`
+    - `id` - ID of the `OrderLine`.
+    - `quantity` - The quantity of the specific lines planned to refund.
+    - `reason` - Reason of the planned refund related to the specific line.
```
Comment (2can) [inline docs/developer/payments.mdx:1014]:
```suggestion
    - `reason` - Reason for the planned refund related to the specific line.
```

File: docs/developer/payments.mdx [hunk 8]
```diff
@@ -852,6 +852,199 @@ sequenceDiagram
...
+
+The mutation accepts below arguments:
+
+- `id` - ID of the transaction which will be used to trigger the refund action.
+- `actionType` - The type of action to be performed on the requested `TransactionItem`. For a refund action, use `REFUND`.
+- `amount` - The amount of the action. If not provided Saleor will use `TransactionItem.chargedAmount`.
+
+#### Refund based on OrderGrantedRefund
+
+When processing a refund based on `OrderGrantedRefund`, there are two steps involved. Firstly, an order must be granted a refund, which defines what should be refunded. This step requires the `MANAGE_ORDERS` permission. Secondly, a refund needs to be requested based on the created order granted refund. This step requires the `HANDLE_PAYMENTS` permission.
```
Comment (2can) [inline docs/developer/payments.mdx:898]:
```suggestion
When processing a refund based on `OrderGrantedRefund`, there are two steps involved. Firstly, an order must be granted a refund, which defines what should be refunded. This step requires the `MANAGE_ORDERS` permission. Secondly, a refund needs to be requested based on the created `OrderGrantedRefund`. This step requires the `HANDLE_PAYMENTS` permission.
```

File: docs/developer/payments.mdx [hunk 9]
```diff
@@ -852,6 +852,199 @@ sequenceDiagram
...
+
+A granted refund contains all the details related to the refund, such as the list of refunded lines, the amount, and the included shipping costs. This is useful when handling refunds based on the products returned by the customer.
+
+The current status of the `orderGrantedRefund` is represented by the `status` field. The possible statuses are:
```
Comment (2can) [inline docs/developer/payments.mdx:902]:
```suggestion
The current status of the `OrderGrantedRefund` is represented by the `status` field. The possible statuses are:
```

File: docs/developer/payments.mdx [hunk 10]
```diff
@@ -852,6 +852,199 @@ sequenceDiagram
...
+    }
+  }
+}
+```
+
+The mutation accepts below arguments:
+
+- `id` - ID of the `OrderGrantedRefund` which should be updated
+- `input`:
+  - `addLines`: Lines that should be added to `OrderGrantedrefund`
```
Comment (2can) [inline docs/developer/payments.mdx:1011]:
```suggestion
  - `addLines`: Lines that should be added to `OrderGrantedRefund`
```

**Review by 2can — CHANGES_REQUESTED** [review]

**Comment by 2can** [comment]
@korycins the following paragraph seems a bit complex as a whole. Maybe some visualization with table(s) or split into some smaller paragraphs would make it  a bit simpler to read 

>The `OrderGrantedRefund` has an impact on the `authorizeStatus`, `chargeStatus`, and `totalBalance`, as it reduces the total value used to calculate the `totalBalance`. For example, if an order has a total of 100 USD and a single `TransactionItem` with a `chargedAmount` of 100 USD, the `totalBalance` would be 0, as `chargedAmount` - `order.total` results in 0. Adding a granted refund with an amount of 10 USD would result in a `totalBalance` of 10, as `chargedAmount` - (`order.total` - `grantedRefund.amount`) gives 10. The `chargeStatus` will be `OVERCHARGED`. If a refund request is made based on the defined granted refund and is successfully processed by the payment app, it will reduce the `chargedAmount`. The `totalBalance` will be 0, as `chargedAmount` (90USD) - (`order.total` - `grantedRefund.amount`) gives 0. The `chargeStatus` will be changed to `FULL`.

**Comment by korycins** [comment]
Thanks @2can  for the review. I applied the requested changes, I also split the mentioned paragraph and added the table.
Let me know what you think!

**Review by 2can — APPROVED** [review]

**Review by Droniu — APPROVED** [review]

**Pull request merged**